### PR TITLE
Fix the cpp-unit plugin to be configuration cache compatible

### PR DESCRIPTION
**Pull request by adammurdoch**
### Context

Also fix a number of Swift and XCTest tests with newer XCode versions.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftApplicationDependenciesIntegrationTest.groovy [hunk 1]
```diff
@@ -19,8 +19,10 @@ package org.gradle.language.swift
 import org.gradle.language.AbstractNativeProductionComponentDependenciesIntegrationTest
 import org.gradle.nativeplatform.fixtures.RequiresInstalledToolChain
 import org.gradle.nativeplatform.fixtures.ToolChainRequirement
+import org.gradle.test.fixtures.file.DoesNotSupportNonAsciiPaths
 
 @RequiresInstalledToolChain(ToolChainRequirement.SWIFTC)
+@DoesNotSupportNonAsciiPaths(reason = "swiftc does not support these paths")
```
Comment (adammurdoch) [inline platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftApplicationDependenciesIntegrationTest.groovy:25]: New versions of the Swift tooling seem to blow up when there are non-ascii paths. For now, ignore this and use ascii paths.

File: platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy [hunk 2]
```diff
@@ -245,8 +245,7 @@ class SwiftIncrementalCompileIntegrationTest extends AbstractInstalledToolChainI
         outputs.recompiledClasses('main', 'sum', 'greeter', 'multiply')
     }
 
-    @RequiresInstalledToolChain(ToolChainRequirement.SWIFTC_4)
-    @ToBeFixedForConfigurationCache
+    @RequiresInstalledToolChain(ToolChainRequirement.SWIFTC_5)
```
Comment (adammurdoch) [inline platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftIncrementalCompileIntegrationTest.groovy:248]: Changed test to use the most recent version of Swift so it is more likely to run.

File: platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftLibraryIntegrationTest.groovy [hunk 3]
```diff
@@ -76,6 +78,7 @@ class SwiftLibraryIntegrationTest extends AbstractInstalledToolChainIntegrationS
         result.assertTasksExecuted(":compileDebugSwift", ":linkDebug", ":assembleDebug")
         file("build/modules/main/debug/${lib.moduleName}.swiftmodule").assertIsFile()
         sharedLibrary("build/lib/main/debug/${lib.moduleName}").assertExists()
+        sharedLibrary("build/lib/main/debug/${lib.moduleName}").assertHasDebugSymbolsFor(lib.sourceFileNames)
```
Comment (adammurdoch) [inline platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/SwiftLibraryIntegrationTest.groovy:81]: Some missing coverage.

File: platforms/native/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/UnexportMainSymbol.java [hunk 4]
```diff
@@ -112,7 +114,7 @@ private void unexportMainSymbol(File object) {
                 throw UncheckedException.throwAsUncheckedException(e);
             }
         } else {
-            getProject().exec(new Action<ExecSpec>() {
+            getExecOperations().exec(new Action<ExecSpec>() {
```
Comment (adammurdoch) [inline platforms/native/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/UnexportMainSymbol.java:117]: to fix CC compatibility.

File: platforms/native/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/NativeBinaryFixture.groovy [hunk 5]
```diff
@@ -121,8 +124,19 @@ class NativeBinaryFixture {
         } else {
             def symbols = binaryInfo.listDebugSymbols()
             def symbolNames = symbols.collect { it.name }
+            def isOlderSwiftc = toolChain.meets(ToolChainRequirement.SWIFTC) && toolChain.version < VersionNumber.version(5, 10)
             sourceFileNames.each { sourceFileName ->
-                assert sourceFileName in symbolNames
+                if (sourceFileName in symbolNames) {
+                    return
+                }
+                if (isOlderSwiftc) {
```
Comment (adammurdoch) [inline platforms/native/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/NativeBinaryFixture.groovy:132]: Swiftc seems to now add the source file names to the debugging information, so look for this first and only look for the object files if the source file names are not present.

File: platforms/native/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/NMToolFixture.groovy [hunk 6]
```diff
@@ -42,16 +42,39 @@ class NMToolFixture {
...
             //
             // Looks like on Windows (MinGW):
             // main T 0000000000401550
-            def splits = line.split(' ')
-            String name = splits[0]
-            char type = splits[1].getChars()[0]
-            return new BinaryInfo.Symbol(name, type, Character.isUpperCase(type))
+            def splits = line.split(' ').toList()
+
+            // The symbol name is the first element, but is not delimited so if it contains spaces, it will span multiple elements
```
Comment (adammurdoch) [inline platforms/native/platform-native/src/testFixtures/groovy/org/gradle/nativeplatform/fixtures/binaryinfo/NMToolFixture.groovy:54]: The format seems to have changed a little bit, so adjust the code for this. Also, handle symbols with spaces in their names.

File: platforms/native/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/xctest/AbstractSwiftXCTestComponentIntegrationTest.groovy [hunk 7]
```diff
@@ -32,6 +32,9 @@ abstract class AbstractSwiftXCTestComponentIntegrationTest extends AbstractSwift
     def setup() {
         // TODO: Temporarily disable XCTests with Swift3 on macOS
         Assume.assumeFalse(OperatingSystem.current().isMacOsX() && toolChain.version.major == 3)
+
+        // Need XCTest available to run these tests
```
Comment (adammurdoch) [inline platforms/native/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/xctest/AbstractSwiftXCTestComponentIntegrationTest.groovy:36]: XCTest is distributed in XCode, not the command-line tools, so skip the tests if XCode is not installed.

File: platforms/native/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/xctest/SwiftXCTestErrorHandlingIntegrationTest.groovy [hunk 8]
```diff
@@ -76,7 +83,10 @@ class SwiftXCTestErrorHandlingIntegrationTest extends AbstractInstalledToolChain
         def testFailure = testExecutionResult.testClass("Gradle Test Run :app:xcTest")
         testFailure.assertTestFailed(EXECUTION_FAILURE, containsText("finished with non-zero exit value"))
         if (OperatingSystem.current().isMacOsX()) {
-            testFailure.assertStderr(containsText("The bundle “AppTest.xctest” couldn’t be loaded because it is damaged or missing necessary resources"))
+            if (toolChain.version < VersionNumber.version(5, 9)) {
+                testFailure.assertStderr(containsText("The bundle “AppTest.xctest” couldn’t be loaded because it is damaged or missing necessary resources"))
+            }
+            // Else, there is no stderr/stdout produced by newer versions
```
Comment (adammurdoch) [inline platforms/native/testing-native/src/integTest/groovy/org/gradle/nativeplatform/test/xctest/SwiftXCTestErrorHandlingIntegrationTest.groovy:89]: The stderr from `xcrun` is going missing on newer swiftc versions, in the same way it's going missing when non-ascii paths are used. I didn't get to the bottom of this, but I think it's ok for now to just make sure the test failed in the expected way.

File: platforms/native/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppUnitTestPlugin.java [hunk 9]
```diff
@@ -138,12 +141,13 @@ private Stream<DefaultCppTestExecutable> getAllTestExecutable() {
                 task.setDescription("Executes C++ unit tests.");
 
                 final InstallExecutable installTask = binary.getInstallTask().get();
-                task.onlyIf("Test executable installation directory exists", element -> binary.getInstallDirectory().get().getAsFile().exists());
+                DirectoryProperty installDirectory = binary.getInstallDirectory();
```
Comment (adammurdoch) [inline platforms/native/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppUnitTestPlugin.java:144]: Fixes for CC compatibility.

File: platforms/native/language-native/src/integTest/groovy/org/gradle/language/cpp/CppSystemHeaderDiscoveryIntegrationTest.groovy [hunk 10]
```diff
@@ -16,28 +16,25 @@
...
 
 class CppSystemHeaderDiscoveryIntegrationTest extends AbstractInstalledToolChainIntegrationSpec {
 
-    @ToBeFixedForConfigurationCache
     def "discovers C and C++ standard library headers"() {
         def outputFile = file("dirs.txt")
         buildFile << """
             plugins { id 'cpp-application' }
             task sysHeaders {
+                def out = file("${outputFile.toURI()}")
```
Comment (adammurdoch) [inline platforms/native/language-native/src/integTest/groovy/org/gradle/language/cpp/CppSystemHeaderDiscoveryIntegrationTest.groovy:29]: Make the test CC compatible.

File: platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/AbstractSwiftIntegrationTest.groovy [hunk 11]
```diff
@@ -101,9 +99,11 @@ abstract class AbstractSwiftIntegrationTest extends AbstractSwiftComponentIntegr
         buildFile << configureTargetMachines("machines.${currentHostOperatingSystemFamilyDsl}", "machines.${currentHostOperatingSystemFamilyDsl}")
         buildFile << """
             task verifyTargetMachines {
+                def targetMachines = ${componentUnderTestDsl}.targetMachines
```
Comment (adammurdoch) [inline platforms/native/language-native/src/integTest/groovy/org/gradle/language/swift/AbstractSwiftIntegrationTest.groovy:102]: Fix test to be CC compatible.

**Review by lkasso — COMMENTED** [review]
Made some updates to the release notes.

File: platforms/native/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppUnitTestPlugin.java (hunk 9 above)
Comment (bamboo) [reply platforms/native/testing-native/src/main/java/org/gradle/nativeplatform/test/cpp/plugins/CppUnitTestPlugin.java:144]: 💡 After extracting the `installDirectory` property, the lambda could have been kept via `SerializableLambdas.spec(...)`

**Review by bamboo — APPROVED** [review]
Looks good 👍

**Pull request merged**